### PR TITLE
Captcha in forms

### PR DESCRIPTION
**Pull request by pettinarip**
## Description

Adds hCaptcha to forms.

TODO
- [x] decide which captcha service to use (using hCaptcha for testing)
- [x] add it to all the forms we have
- [x] test and fix possible cors issues
- [x] support mobile
- [x] validate token in serverside

#198

### CODE REVIEW

File: src/components/forms/AcademicGrantsForm.tsx [hunk 1]
```diff
@@ -115,1388 +119,1400 @@ export const AcademicGrantsForm: FC = () => {
...
+            {errors?.additionalInfo?.type === 'maxLength' && (
+              <Box mt={1}>
+                <PageText as='small' fontSize='helpText' color='red.500'>
+                  Additional info cannot exceed 32768 characters.
+                </PageText>
+              </Box>
+            )}
+          </FormControl>
+
+          <Center mb={12}>
```
Comment (pettinarip) [inline src/components/forms/AcademicGrantsForm.tsx:1476]: The only significant change in the html

Comment (pettinarip) [reply src/components/forms/AcademicGrantsForm.tsx:1476]: Note: only changes in the forms html are the inclusion of `<FormProvider>` and `<Captcha />`. You can ignore the rest of the diff, nothing has changed.

File: additional.d.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import { Fields, Files } from 'formidable';
```
[the remark this answers is not in the document]
Comment (pettinarip) [reply additional.d.ts:1]: Basically to extend the `NextApiRequest` interface from Next with custom added fields. Check the `multipartyParse` middleware, I'm setting these fields there to later be used by other middlewares (or the final handler) in the chain.

File: src/components/forms/AcademicGrantsForm.tsx [hunk 3]
```diff
@@ -104,1363 +106,1375 @@ export const AcademicGrantsForm: FC = () => {
...
-                  <PageText as='small' fontSize='helpText' color='red.500'>
-                    First name is required.
-                  </PageText>
-                </Box>
-              )}
-              {errors?.firstName?.type === 'maxLength' && (
-                <Box mt={1}>
-                  <PageText as='small' fontSize='helpText' color='red.500'>
-                    First name cannot exceed 40 characters.
+      <FormProvider {...methods}>
```
[the remark this answers is not in the document]
Comment (pettinarip) [reply src/components/forms/AcademicGrantsForm.tsx:109]: In order to use `useFormContext` from inside the Captcha component you need a provider wrapping the form you are working on. With that, we can isolate all the captcha-form logic in its component and access to the form's methods.

From the consumer, you are able to just call the component and forget about any other logic:
```jsx
<FormProvider {...methods}>
  ...
  
  <Captcha />
</FormProvider>
```

[the remark this answers is not in the document]
Comment (nhsz) [reply src/components/forms/AcademicGrantsForm.tsx:109]: Great!

**Comment by nhsz** [comment]
@pettinarip LGTM, great work! 🎉 Haven't approved yet because maybe the middleware logic can be abstracted using Next's [middleware](https://nextjs.org/docs/middleware). In case you already check that I'll approve.

**Comment by nhsz** [comment]
> middleware

Agree. You can apply the middleware to certain routes only https://nextjs.org/docs/middleware#execution-order, so maybe nesting the `multipartyParse` middleware inside that dir should execute it for that route only 🤔

**Comment by pettinarip** [comment]
> maybe nesting the multipartyParse middleware inside that dir should execute it for that route only thinking

You mean having something like this:
```
- /pages
    _middleware.ts # check captcha tokens
    devcon-grants.tsx
    - /project-grants
      _middleware.ts # apply multiparty parsing
      index.tsx
```
The problem with that is that I need to apply the parsing before the captcha check in this case.

I think I will go with just one `_middleware.ts` at the root level and add a condition to apply the multiparty parsing to a specific route.

**Comment by pettinarip** [comment]
I'm having some troubles querying `req.body` on a `NextRequest`. It seems that is it not being sent to this type of request.
When I do `await req.json()` I receive an error `TypeError: invalid json body reason: Unexpected end of JSON input` due to `req.body` being empty 🤷🏼 

I've upgraded Next to `12.1.0` which has this PR https://github.com/vercel/next.js/pull/34294 but it is not working for me.

Then `formidable` is using the `fs` module which is not allowed to be used in Edge functions https://nextjs.org/docs/api-reference/edge-runtime#unsupported-apis

Given these conditions I would say to skip this refactor for now.

**Review by nhsz — APPROVED** [review]

**Pull request merged**